arangod/Aql/OptimizerRules.cpp

```diff
@@ -4093,12 +4220,16 @@ class RestrictToSingleShardChecker final : public WalkerWorker<ExecutionNode> {
       case EN::REPLACE:
       case EN::UPDATE:
       case EN::REMOVE: {
-        auto collection = ExecutionNode::castTo<ModificationNode const*>(en)->collection();
+        auto collection =
+            ExecutionNode::castTo<ModificationNode const*>(en)->collection();
         std::string shardId = getSingleShardId(_plan, en, collection);
-        if (shardId.empty()) {
-          _shardsUsed[collection].clear();
+        if (shardId.empty() && _shardsUsed[collection].empty()) {
           _shardsUsed[collection].emplace("all");
-        } else {
+        } else if (!shardId.empty()) {
+          if (1 == _shardsUsed[collection].size() &&
+              "all" == *_shardsUsed[collection].begin()) {
+            _shardsUsed[collection].clear();
```

Comment:
What happens by the way, if we find a FILTER on a collection after an INSERT/UPDATE/REMOVE, e.g.
```
FOR doc IN collection
  UPDATE doc WITH ... IN collection
  FILTER doc.shardKey == @whatever
  RETURN doc
```
AFAIK this must do the updates on all shards, and then only return the documents from one shard. Is this covered?

Reply:
This is a contrived example, however, we need to support all types of queries.